File: olimpiada_informatyczna/reprezentacje_roznicowe/rep.cpp
/*  Jan Zakrzewski
    XXIV Olimpiada Informatyczna, Etap I
    Zadanie Reprezentacje różnicowe (rep)
    Rozwiązenie w złożoności log^2 X + n log log X */

#include <bits/stdc++.h>
using namespace std;

int binsearch(vector<tuple<int,int,int>> arr, int x) {
    if(get<0>(arr[0]) >= x) return 0;
    int a = 0;
    int b = arr.size();
    while(b - a > 1) {
        int mid = (a + b) / 2;
        if(get<0>(arr[mid]) >= x) b = mid;
        else a = mid;
    }
    return b;
}

constexpr int X = 1e9;
constexpr int N = 70;
int a[N], t, n;

bool r_seen[N*N+20];
int r;

vector<tuple<int, int, int>> uneasy;

int main() {

    ios_base::sync_with_stdio(0);
    cin.tie(0), cout.tie(0);

    a[0] = 1, a[1] = 2;

    r = 2, r_seen[1] = true;
    for(int i = 2; i < N*N+20; ++i) r_seen[i] = false;
    n = N;
    for(int i = 2; i < N; ++i) {
        if(i % 2 == 0) a[i] = a[i-1] * 2;
        else a[i] = a[i-1] + r;
        if(a[i] > INT32_MAX / 2 - 20) { n = i; break; }
        for(int j = 0; j < i; ++j) {
            if(a[i] - a[j] < N*N+20)
                r_seen[a[i] - a[j]] = true;
        }
        while(r < N*N+20 && r_seen[r]) r++;
    }

    for(int i = 1; i < n; ++i) {
        for(int j = 0; j < i; ++j) {
            int x = a[i] - a[j];
            if(x <= X && (i - j > 1 || j % 2 == 1))
                uneasy.push_back(make_tuple(x, i, j));
        }
    }
    sort(uneasy.begin(), uneasy.end());

    function<pair<int, int>(int)> repr = [&](int x)->pair<int, int> {
        int i = binsearch(uneasy, x);
        if(i < uneasy.size() && get<0>(uneasy[i]) == x)
            return make_pair(get<1>(uneasy[i]), get<2>(uneasy[i]));
        int q = 2 * (x - i - 1);
        int p = q + 1;
        return make_pair(p, q);
    };

    cin >> t;
    while(t--) {
        int x;
        cin >> x;
        auto elm = repr(x);
        int p = elm.first, q = elm.second;
        cout << p+1 << " " << q+1 << "\n";        
    }
    
    return 0;
}

File: olimpiada_informatyczna/reprezentacje_roznicowe/rep.py
''' Jan Zakrzewski
    XXIV Olimpiada Informatyczna, Etap I
    Zadanie Reprezentacje różnicowe (rep)
    Rozwiązenie w złożoności log^2 X + n log log X '''

def binsearch(arr, x):
    '''Returns the smallest `i` for which `arr[i][0] >= x` or `len(arr)` if there is no such `i`.'''
    if arr[0][0] >= x: return 0
    a, b = 0, len(arr)
    while b - a > 1:
        mid = (a + b) // 2
        if arr[mid][0] >= x: b = mid
        else: a = mid
        #print(a, b, arr[a][0] <= x and arr[b][0] > x)
    return b

X = 10**9
N = 70 # assuming a[N] - a[i] > X for any 0<=i<N
a = [1, 2]

r_seen = [False for _ in range(N * N)]
r_seen[1] = True
r = 2
for i in range(2, N):
    if i % 2 == 0: a.append(a[i-1] * 2)
    else: a.append(a[i-1] + r)
    for j in range(0, i):
        if a[i] - a[j] < N * N:
            r_seen[a[i] - a[j]] = True
    while r < N * N and r_seen[r]:
        r += 1
del r, r_seen

uneasy = []

for i in range(1, N):
    for j in range(0, i):
        x = a[i] - a[j]
        p, q = i, j
        if x <= X and (p - q > 1 or q % 2 == 1):
            uneasy.append((x, p, q))
uneasy = sorted(uneasy)

def repr(x):
    i = binsearch(uneasy, x)
    if i < len(uneasy) and uneasy[i][0] == x:
        return uneasy[i][1], uneasy[i][2]
    q = 2 * (x - i - 1)
    p = q + 1
    return p, q

t = int(input())
for _ in range(t):
    x = int(input())
    p, q = repr(x)
    print('{} {}'.format(p+1, q+1))
